### DISCUSSION TITLE
Removing the {{Draft}} macro

### DISCUSSION THREAD

**Discussion by wbamberg** [mdn/mdn-community#583] [Content] [answered]
Some time ago we triaged KS macros in the JS to try to decide which ones we wanted to keep: https://github.com/mdn/yari/issues/2846#issuecomment-593693054 .

Of the ones we wanted to remove it looks like only `{{Draft}}` is left. I have noticed people removing this here and there, but wonder if we have consensus to remove it all across the board.

**Rationale:**

* We shouldn't publish content on MDN that's not useful, so shouldn't use `{{Draft}}` to indicate that. We do have lots of incomplete content, only a small fraction of which is marked up with `{{Draft}}`. We should deal with this in the usual way by responding to issues and plan projects to fix it.

* On the other hand almost all content is draft in that it could potentially be improved or expanded, so `{{Draft}}` doesn't add anything.

* It's also not clear what use `{{Draft}}` is supposed to be for readers - if the content is incomplete that's obvious, if some of it is wrong it doesn't tell people which bits or how.  Often `{{Draft}}` is left on a page long after it's already been made useful enough, so it only serves to make people anxious about the content, without giving them enough information to do something with that anxiety.

**Comment by teoli2003** [comment]
I'm 100% for removing `{{Draft}}`. If we have consensus, I'm for finding/replacing all occurrences and removing them and any mention of the macro in our meta documentation and plain deleting the macro.

With the PR system, non-ready documents don't get published. It is a remnant of the wiki era.

**Reply by jpmedley** [reply]
I completely agree.

**Reply by hamishwillee** [reply]
If the macro is of any use to us as editors then you might leave it in but simply change the implementation to have no rendering.

**Reply by hamishwillee** [reply]
Interesting list - there are lots of `deprecated_inline` inline's around still. What were we planning on replacing them with?

**Reply by wbamberg** [reply]
> there are lots of deprecated_inline inline's around still.

The linked issue was scoped to the JS docs.

**Reply by tannerdolby** [reply]
To add on this, if the consensus is to move forward with removing any and all `{{Draft}}` usage in `mdn/content` and `mdn/yari`. With both repos up-to-date. Here is the breakdown of occurrences:

Edit: (Before all `{{Draft|draft}}` calls were removed):
- 279 occurrences of `{{draft}}` across 279 files in mdn/content
- 42 occurrences of `{{Draft}}` across 41 files in mdn/content
- 4 occurrences of `{{Draft}}` across 1 file in Yari (within Draft.test.js)
- 312 combined occurrences of `{{Draft}}` or `{{draft}}` in mdn/translated-content

I also noticed that some templates like `svg-in-opentype/index.md` still have `Draft` defined in the `tags` key inside frontmatter data. Would this sort of reference to `Draft` be removed as well?

**Reply by wbamberg** [reply]
There are another 279 for `{{draft}}` :) but that's not too many still. Probably lots in translated content too.

> I also noticed that some templates like svg-in-opentype/index.md still have Draft defined in the tags key inside frontmatter data. Would this sort of reference to Draft be removed as well?

This is a bit different, but I would love to clean up (read: delete most of) our tags as well -> https://github.com/mdn/content/discussions/5805.

**Reply by estelle** [reply]
I do think we need some form of “Under Construction” so people can build out an entire API and have placeholder pages with visitors knowing “this page is not yet built”.

Example: Right now I am doing the attribute section one page at a time so one PR at a time. The roles i did all at once and it was too large of a PR to review, so included “draft”.

The one page per PR worked because I am doing it solo, but if i wanted help building I would build out a template for all the interfaces of an API with “draft” so that others could add examples / descriptions based on a shell outline. If we don’t have some form of “under construction” we won’t be as able to group build.

**Reply by estelle** [reply]
When i create pages, i literally put draft and needs content in my bash file: 
```bash
cat > ${newFileName}/index.md <<- EOM
---
title: 'ARIA: ${newFileName} attribute'
slug: Web/Accessibility/ARIA/Attributes/${newFileName}
tags: 
$tags
  - Accessibility
  - ARIA
  - ARIA attribute
  - ARIA property
  - ${newFileName}
  - NeedsContent
---
{{draft}}
```

**Reply by wbamberg** [reply]
`{{Draft}}` was needed in the Wiki because there was no real way to write a first draft of something, get feedback and iterate, without publishing it. But now we are in GitHub that doesn't seem like a necessity.

> The one page per PR worked because I am doing it solo, but if i wanted help building I would build out a template for all the interfaces of an API with “draft” so that others could add examples / descriptions based on a shell outline. If we don’t have some form of “under construction” we won’t be as able to group build.

I appreciate that having empty placeholder pages is one way to do this, but it's not the only way. Couldn't we coordinate large shared projects using a tracking issue?

I think it's better to separate the tools we have as writers from the experiences of people who are just reading the docs, and don't care about the internals, and as far as possible we should present people with finished pages. If it take some iterations to get there we can do that in review.

**Reply by hamishwillee** [reply]
I don't have a problem with Under Construction for that purpose - you're running a project and you need tools for that.

I also quite like the idea of using template placeholders for API reference with a "Please populate me" header  - in many ways that is better than a 404 links as we have now.

What I don't like is the half way point as Draft is now - something that may or may not be finished, but generally there is no way to know what is expected of the reader. 
I think we should clean up what there is now and then allow Draft or even better "under construction" but only for use as part of projects.

**Reply by hamishwillee** [reply]
> I think it's better to separate the tools we have as writers from the experiences of people who are just reading the docs, and don't care about the internals, and as far as possible we should present people with finished pages. If it take some iterations to get there we can do that in review.

The rule that everything is expected to be good at the point of publish is very consistent, and simple for us as authors and reviewers to understand. Also good for readers - "this should be accurate" is a great simple rule.

That said, under construction headers or whatever have the benefit you can use them to encourage contribution. They also have value if BCD is complete, as the user gets the compat and spec data already.

**Reply by jpmedley** [reply]
>I do think we need some form of “Under Construction” so people can build out an entire API and have placeholder pages with visitors knowing “this page is not yet built”.

If you can do it without affecting SEO. The cumulative affect of search results that don't contain useful information would not be good for the site.

**Reply by caugner** [reply]
FWIW As of today (2022-03-10), the macro is still used 14 times in content and 44 times in translated-content (I checked this in the context of https://github.com/mdn/yari/pull/4943).

**Comment by estelle** [comment]
Still need to get the a11y roles merged... those are the last drafts.

**Comment by Rumyra** [comment]
I noticed when doing release notes this month that they were still in the FF rel notes (I removed them in new files I created) - now it's 5 on my count (all a11y) - so we're close!

**Comment by estelle** [comment, marked as answer]
I don't see {{Draft}} in my local code base. Can we 🎉 ?

**Reply by tannerdolby** [reply]
With [#385](https://github.com/mdn/translated-content/pull/3085#issuecomment-1157641740) merged and the other separate PRs associated with {{Draft}} removal complete. I believe it is time to 🎉 !

**Reply by teoli2003** [reply]
In today's Yari, `{{draft}}`is definitively gone!
